docs(framework) Add Flower Network Communication reference (#4805)
adap/flower: framework/docs/source/ref-flower-network-communication.rst

In isolation mode ``process``, additional network connections are necessary to allow the external process running ServerApp or ClientApp to communicate with the long-running SuperLink or SuperNode:

- **flwr-serverapp to SuperLink (ServerAppIO API)**: The process running the ``ServerApp``, ``flwr-serveapp``, acts as a gRPC client and connects to the SuperLink's ServerAppIO API.
  This connection enables the ``flwr-serverapp`` process to pull the necessary inputs to execute the ``ServerApp``.
  It also allows the ``ServerApp``, once running, to do typical things like sending/receiving messages to/from available SuperNodes (via the SuperLink).
- **flwr-clientapp to SuperNode (ClientAppIO API)**: The process running the ``ClientApp``, ``flwr-clientapp``, acts as a gRPC client and connects to the SuperNode's ClientAppIO API.
  This connection enables the ``flwr-clientapp`` process to pull the necessary details (e.g., FAB file) to execute the ``ClientApp``, execute the ``ClientApp`` (e.g., local model training) and return the execution results (e.g., locally update model parameters) to the SuperNode.

.. note::

>   In the current version of Flower, both of these connections are insecure because Flower expects SuperLink/SuperNode and ``flwr-serverapp`` / ``flwr-clientapp`` to be run in the same enviornment.

Author: > 

>

SuperLink or SuperNode:

- **flwr-serverapp to SuperLink (ServerAppIO API)**: The process running the
  ``ServerApp``, ``flwr-serveapp``, acts as a gRPC client and connects to the
  SuperLink's ServerAppIO API. This connection enables the ``flwr-serverapp`` process to
  pull the necessary inputs to execute the ``ServerApp``. It also allows the
  ``ServerApp``, once running, to do typical things like sending/receiving messages
  to/from available SuperNodes (via the SuperLink).
- **flwr-clientapp to SuperNode (ClientAppIO API)**: The process running the
  ``ClientApp``, ``flwr-clientapp``, acts as a gRPC client and connects to the
  SuperNode's ClientAppIO API. This connection enables the ``flwr-clientapp`` process to
  pull the necessary details (e.g., FAB file) to execute the ``ClientApp``, execute the
  ``ClientApp`` (e.g., local model training) and return the execution results (e.g.,
  locally update model parameters) to the SuperNode.

.. note::

    In the current version of Flower, both of these connections are insecure because
    Flower expects SuperLink/SuperNode and ``flwr-serverapp`` / ``flwr-clientapp`` to be

>     run in the same enviornment. ``flwr-serverapp`` / ``flwr-clientapp`` and

Author: ```suggestion
    run in the same network. ``flwr-serverapp`` / ``flwr-clientapp`` and
```

    SuperLink/SuperNode should never communicate over untrusted networks (e.g., public
    internet).

User authentication
~~~~~~~~~~~~~~~~~~~

When user authentication is enabled, Flower uses an OIDC-compatible server to
authenticate requests:

- **SuperLink to OIDC server**: A SuperLink can optionally be configured to only allow
  authenticated users to interact with it. In this setting, the Flower SuperLink acts as
  a REST client to the OIDC-compatible server.

Application-specific connections
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Users who write Flower Apps (``ServerApp`` and ``ClientApp``) can also make additional
network requests. This is, strictly speaking, not part of Flower as a Federated AI
Platform. It is a decision of (a) the user about what kinds of third-party systems their
Flower App should connect to and (b) the system administrator about what kinds of
connections they want to allow.

Typical examples include:

> - **ServerApp to Database**: ``ServerApp`` instances might want to access the data to
>   perform the action they have been designed for (e.g. train locally a model, run a DB
>   query). How this connection is established depends on what storage technology used at
>   the client side. Note that in the diagram above, we show two representative
>   connections to DBs in Client-A and Client-B. Your DB connection(s) may likely be
>   different to the illustration above.

Author: ```suggestion
- **ServerApp to Database**: ``ServerApp`` instances might want to access the data to
  perform the action they have been designed for (e.g. evaluate a model on some data after aggregation). How this connection is established depends on what storage technology used at
  the client side. Note that in the diagram above we omitted showing a DB connected to the ServerApp components.
```

> - **ClientApp to Database**: ``ClientApp`` instances typically need to be able to access
>   the data to perform the action they have been designed for (e.g. train locally a
>   model, run a DB query). How this connection is established depends on what storage
>   technology used at the client side. Note that in the diagram above, we show two
>   representative connections to DBs in Client-A and Client-B. Your DB connection(s) may
>   likely be different to the illustration above.

Author: ```suggestion
- **ClientApp to Database**: ``ClientApp`` instances typically need to be able to access
  the data to perform the action they have been designed for (e.g. train locally a
  model, run a DB query). How this connection is established depends on what storage
  technology is used at the client side. Note that in the diagram above, we show two
  representative connections to DBs in Client-A and Client-B. Your DB connection(s) may
  likely be different to the illustration above.
```
